src/engine/IndexScan.cpp

```diff
@@ -595,75 +664,138 @@ struct IndexScan::SharedGeneratorState {
-                           innerState->doneFetching_);
-    }
-    // Make a defensive copy of the values to avoid modification during
-    // iteration when yielding.
-    auto copy = std::move(prefetchedValues);
-    // Moving out does not necessarily clear the values, so we do it explicitly.
-    prefetchedValues.clear();
-    for (auto& value : copy) {
-      co_yield value;
-    }
-  }
+  auto range = ad_utility::InputRangeFromGetCallable{
+      [state = std::move(innerState),
+       buffer = SharedGeneratorState::PrefetchStorage{}]() mutable
+      -> std::optional<Result::IdTableVocabPair> {
+        // Handle UNDEF case: pass through remaining input
+        if (state->hasUndef()) {
+          if (!state->iterator_.has_value()) {
+            state->iterator_ = state->generator_.begin();
+          }
+          auto& it = state->iterator_.value();
+          if (it == state->generator_.end()) {
+            return std::nullopt;
+          }
+          auto result = std::move(*it);
+          ++it;
+          return result;
+        }
+
+        auto& prefetched = state->prefetchedValues_;
+        // Helper to fill buffer from prefetched values
+        auto fillBuffer = [&]() -> bool {
+          while (buffer.empty()) {
+            if (prefetched.empty()) {
+              if (state->doneFetching_) {
+                return false;
+              }
+              state->fetch();
+              AD_CORRECTNESS_CHECK(!prefetched.empty() || state->doneFetching_);
+              if (prefetched.empty()) {
+                return false;
+              }
+            }
+            buffer = std::move(prefetched);
+            prefetched.clear();
+          }
+          return true;
+        };
+
+        if (!fillBuffer()) {
+          return std::nullopt;
+        }
+        auto result = std::move(buffer.front());
+        buffer.erase(buffer.begin());
+        return result;
+      }};
+  return Result::LazyResult{std::move(range)};
 }
 
 // _____________________________________________________________________________
-Result::Generator IndexScan::createPrefilteredIndexScanSide(
+Result::LazyResult IndexScan::createPrefilteredIndexScanSide(
     std::shared_ptr<SharedGeneratorState> innerState) {
   if (innerState->hasUndef()) {
-    for (auto& pair : chunkedIndexScan()) {
-      co_yield pair;
-    }
-    co_return;
-  }
-  LazyScanMetadata metadata;
-  auto& pendingBlocks = innerState->pendingBlocks_;
-  while (true) {
-    if (pendingBlocks.empty()) {
-      if (innerState->doneFetching_) {
-        metadata.numBlocksAll_ = innerState->metaBlocks_.sizeBlockMetadata_;
-        updateRuntimeInfoForLazyScan(metadata);
-        co_return;
-      }
-      innerState->fetch();
-    }
-    auto scan = getLazyScan(std::move(pendingBlocks));
-    AD_CORRECTNESS_CHECK(pendingBlocks.empty());
-    for (IdTable& idTable : scan) {
-      co_yield {std::move(idTable), LocalVocab{}};
-    }
-    metadata.aggregate(scan.details());
```

Comment:
This also can be refactored using the `LoopControl` facilities using the `YieldAll` etc.